text/0000-assoc-default-groups.md

```diff
@@ -698,6 +739,199 @@ to change this behaviour. Presumably, this is the case because Haskell
 preserves parametricity and lacks specialization,
 wherefore `default { .. }` might not carry its weight.
 
+## Idris
+
+[idris_interface]: http://docs.idris-lang.org/en/latest/tutorial/interfaces.html
+
+Idris has a concept it calls [`interface`s][idris_interface].
+These resemble type classes in Haskell, and by extension traits in Rust.
+However, unlike Haskell and Rust, these `interface`s are incoherent and will
+permit multiple implementations of the same interface.
+
+Since Idris is language with full spectrum dependent types,
+it does not distinguish between terms and types, instead, types are terms.
+Therefore, there is really not a distinct concept called "associated type".
+However, an `interface` may require certain definitions to be provided
+and this includes types. For example, we may write:
+
+```idris
+interface Iterator self where
+    item : Type
+    next : self -> Maybe (self, item)
+
+implementation Iterator (List a) where
+    item = a
+    next [] = Nothing
+    next (x :: xs) = Just (xs, x)
+```
+
+Like in Haskell, in Idris, a function or value in an interface may be given a
+default definition. For example, the following is a valid program:
+
+```idris
+interface Foo x where
+    bar : Type
+    bar = Bool
+
+    baz : x -> bar
+
+implementation Foo Int where
+    baz x = x == 0
+```
+
+However, if we provide a default for `baz` in the `interface` which assumes
+the default value `Bool` of `bar`, as with the following example:
+
+```idris
+interface Foo x where
+    bar : Type
+    bar = Bool
+
+    baz : x -> bar
+    baz _ = True
+```
+
+then we run into an error:
+
+```
+Type checking .\foo.idr
+foo.idr:6:13-16:
+  |
+6 |     baz _ = True
+  |             ~~~~
+When checking right hand side of Main.default#baz with expected type
+        bar x _
+
+Type mismatch between
+        Bool (Type of True)
+and
+        bar x _ (Expected type)
+```
+
+The behaviour here is exactly as in Haskell and as proposed in this RFC.
+
+## C++
+
+In C++, it is possible to provide associated types and specialize them as well.
+This is shown in the following example:
+
+```cpp
+#include <iostream>
+#include <string>
+
+template<typename T> struct wrap {};
+
+template<typename T> struct foo { // Unspecialized.
+    typedef int bar;
+
+    bar make_a_bar() { return 0; };
+};
+
+template<typename T> struct foo<wrap<T>> { // Partial specialization.
+    typedef std::string bar;
+
+    bar make_a_bar() { return std::string("hello world"); };
+};
+
+int main() {
+    foo<void> a_foo;
+    std::cout << a_foo.make_a_bar() << std::endl;
+
+    foo<wrap<void>> b_foo;
+    std::cout << b_foo.make_a_bar() << std::endl;
+}
+```
+
+One thing to note here is that C++ allows us to assume in both the unspecialized
+variant of `foo` as well as the specialized version that `bar` is the underlying
+type we said it was in the `typedef`. This is unlike the default in this RFC
+but the same as when we use `default { .. }`.
```

Comment:
I'd rewrite this comment quite heavily, to something along the following lines:

> You will note that C++ allows us to assume in both the base template class, as well as the specialization, that `bar` is equal to the underlying type. This is because one cannot specialize any part of a class without specializing the whole of it. It's equivalent to one atomic `default { ... }` block.